File: pets24x7_new/scripts/strip_synthetic_ratings.py
"""
Strip the invented ratings that older builds of build_data.py baked into
data/*.json (and the PETS_FEATURED block of pets-data.js).

Those builds filled gaps in the scraped CSVs with made-up numbers:
  - rating 4.4 whenever the "Review" column had no "Rated X out of 5"
  - review_count = 18 + (digit_sum(cid) * 7919) % 462 whenever "Rating count"
    was empty (digit_sum of 0 counts as 31)
Both were published as "Rated 4.4/5 on Google from 237 reviews" and as
AggregateRating JSON-LD. The formula is deterministic, so a baked value can be
recognised exactly:
  - review_count equal to the formula for its CID  -> 0 (no count known)
  - rating 4.4 on such a row                        -> null (no rating known)
A rating other than 4.4 came from the source ("Rated 4.8 out of 5") and is
kept; with no count behind it every template hides it anyway. Rows are
otherwise untouched, so the output of a second run equals its input.

  python scripts/strip_synthetic_ratings.py            # dry run, prints counts
  python scripts/strip_synthetic_ratings.py --apply    # rewrite changed files
  python scripts/strip_synthetic_ratings.py --apply --no-index   # data/ only
"""
import argparse
import json
import re
import sys
from pathlib import Path

ROOT = Path(__file__).resolve().parent.parent
DATA_DIR = ROOT / "data"
INDEX_FILE = ROOT / "pets-data.js"

DEFAULT_RATING = 4.4


def synthetic_count(cid: str) -> int:
    """The exact count the old build_data.py derived from a CID."""
    seed = sum(int(ch) for ch in cid if ch.isdigit()) or 31
    return 18 + (seed * 7919) % 462


def strip_row(r: dict, stats: dict) -> bool:
    """Clear the invented values on one listing. True if it changed."""
    # Any CID, not only a numeric one: the old build summed the digits of a
    # spreadsheet-mangled "6.47E+18" too, so those rows carry the same fake.
    cid = str(r.get("google_cid") or "")
    if not cid:
        return False
    try:
        count = int(r.get("review_count") or 0)
    except (TypeError, ValueError):
        return False
    if count <= 0 or count != synthetic_count(cid):
        return False
    r["review_count"] = 0
    stats["count"] += 1
    if r.get("rating") == DEFAULT_RATING:
        r["rating"] = None
        stats["rating"] += 1
    else:
        stats["real_rating_kept"] += 1
    return True


def dump_like(original_text: str, value) -> str:
    """Re-serialise in the file's own layout: build_data.py writes compact
    JSON, the API writes imported_listings.json indented."""
    if original_text.lstrip().startswith("[\n") or "\n  " in original_text[:200]:
        return json.dumps(value, ensure_ascii=False, indent=2)
    return json.dumps(value, ensure_ascii=False, separators=(",", ":"))


def process_data(apply: bool, stats: dict) -> None:
    for fp in sorted(DATA_DIR.glob("*.json")):
        text = fp.read_text(encoding="utf-8")
        try:
            rows = json.loads(text)
        except ValueError:
            print(f"[skip] {fp.name}: not JSON", file=sys.stderr)
            continue
        if not isinstance(rows, list):
            continue
        stats["files"] += 1
        stats["rows"] += len(rows)
        changed = sum(1 for r in rows if isinstance(r, dict) and strip_row(r, stats))
        if changed:
            stats["files_changed"] += 1
            if apply:
                fp.write_text(dump_like(text, rows), encoding="utf-8")


# Decoded as one JSON value after the "=", not matched up to the first "];":
# a featured business name can contain "];".
FEATURED_RE = re.compile(r"\bwindow\.PETS_FEATURED\s*=\s*")


def process_index(apply: bool, stats: dict) -> None:
    """pets-data.js carries full rows for the home page's featured strip. Its
    PETS_INDEX top_rating is read by no page, so only the featured rows are
    rewritten; the next build_data.py run recomputes the whole file."""
    if not INDEX_FILE.exists():
        return
    text = INDEX_FILE.read_text(encoding="utf-8")
    m = FEATURED_RE.search(text)
    if not m:
        print(f"[warn] {INDEX_FILE.name}: PETS_FEATURED not found", file=sys.stderr)
        return
    start = m.end()
    try:
        rows, end = json.JSONDecoder().raw_decode(text, start)
    except ValueError as err:
        print(f"[warn] {INDEX_FILE.name}: PETS_FEATURED is not valid JSON ({err})", file=sys.stderr)
        return
    if not isinstance(rows, list):
        print(f"[warn] {INDEX_FILE.name}: PETS_FEATURED is not a list", file=sys.stderr)
        return
    before = dict(stats)
    changed = sum(1 for r in rows if strip_row(r, stats))
    stats["featured_changed"] = changed
    # Featured rows are not listings on their own; keep them out of the totals.
    for k in ("count", "rating", "real_rating_kept"):
        stats[k] = before[k]
    if changed and apply:
        new = json.dumps(rows, ensure_ascii=False, separators=(",", ":"))
        INDEX_FILE.write_text(text[:start] + new + text[end:], encoding="utf-8")


def main():
    ap = argparse.ArgumentParser(description=__doc__, formatter_class=argparse.RawDescriptionHelpFormatter)
    ap.add_argument("--apply", action="store_true", help="write the changes (default: dry run)")
    ap.add_argument("--no-index", action="store_true", help="leave pets-data.js alone")
    args = ap.parse_args()

    stats = {"files": 0, "files_changed": 0, "rows": 0, "count": 0, "rating": 0,
             "real_rating_kept": 0, "featured_changed": 0}
    process_data(args.apply, stats)
    if not args.no_index:
        process_index(args.apply, stats)

    mode = "applied" if args.apply else "dry run (use --apply to write)"
    print(f"[{mode}]")
    print(f"  data files scanned:        {stats['files']:>7}  ({stats['rows']:,} rows)")
    print(f"  data files with changes:   {stats['files_changed']:>7}")
    print(f"  invented review counts -> 0:    {stats['count']:>7}")
    print(f"  default 4.4 ratings -> null:    {stats['rating']:>7}")
    print(f"  source ratings kept (count 0):  {stats['real_rating_kept']:>7}")
    if not args.no_index:
        print(f"  pets-data.js featured rows fixed: {stats['featured_changed']:>5}")


if __name__ == "__main__":
    main()

File: pets24x7_api/scripts/strip-synthetic-ratings.mjs
// One-off clean-up of the invented ratings in the listings table.
//
//   node scripts/strip-synthetic-ratings.mjs            # dry run: counts only
//   node scripts/strip-synthetic-ratings.mjs --apply    # write the fixes
//   node scripts/strip-synthetic-ratings.mjs --export path/to/listings.json
//
// Older builds of pets24x7_new/build_data.py filled gaps in the scraped CSVs
// with made-up numbers, and import-mysql.mjs loaded them into this table:
//   - reviewCount = 18 + (digitsum(googleCid) * 7919) % 462 when the CSV had no
//     count (a digit sum of 0 counts as 31)
//   - rating 4.4 when the CSV had no "Rated X out of 5"
// The formula is deterministic, so a baked count is recognised exactly; it
// becomes 0 (no count known), and a 4.4 on such a row becomes 0 (rating is a
// non-null Float, 0 means unknown). A rating other than 4.4 came from the source
// and is kept; with no count behind it the API and the pages hide it
// (shownRating / build_pages.has_rating). The same rule as
// pets24x7_new/scripts/strip_synthetic_ratings.py for the JSON files.
//
// clean-listing-cids.mjs blanks spreadsheet-mangled CIDs ("6.60859E+18"). The
// old build had summed the digits of those too, so once the CID is gone the
// row can only be recognised through its original CID in export/listings.json
// (the file import-mysql.mjs loaded). Pass --export to point elsewhere; without
// the file those rows are reported and left alone.
//
// A vendor- or admin-edited row with a real count matches only by coincidence
// of the exact formula. Safe to re-run: the second run finds nothing.

import { readFile } from 'node:fs/promises';
import path from 'node:path';
import { fileURLToPath } from 'node:url';

import { PrismaClient } from '@prisma/client';

const APPLY = process.argv.includes('--apply');
const exportArg = process.argv.indexOf('--export');
const EXPORT_FILE =
  exportArg > 0 && process.argv[exportArg + 1]
    ? path.resolve(process.argv[exportArg + 1])
    : path.resolve(path.dirname(fileURLToPath(import.meta.url)), '../export/listings.json');
const DEFAULT_RATING = 4.4;
const BATCH = 500;
const prisma = new PrismaClient();

const syntheticCount = (cid) => {
  let seed = 0;
  for (const ch of String(cid)) if (ch >= '0' && ch <= '9') seed += Number(ch);
  return 18 + ((seed || 31) * 7919) % 462;
};

// digitsum(googleCid) in SQL: each digit times the number of times it occurs.
const DIGITSUM = Array.from({ length: 9 }, (_, i) => i + 1)
  .map((d) => `${d}*(CHAR_LENGTH(googleCid)-CHAR_LENGTH(REPLACE(googleCid,'${d}','')))`)
  .join(' + ');
const MATCH = `googleCid IS NOT NULL AND googleCid <> '' AND reviewCount = 18 + MOD(IF((${DIGITSUM}) = 0, 31, (${DIGITSUM})) * 7919, 462)`;

async function main() {
  const target = (process.env.DATABASE_URL ?? '').replace(/\/\/[^@]*@/, '//***@');
  console.log(`${APPLY ? 'APPLY' : 'DRY RUN'} against ${target}`);

  // 1. Rows that still carry their CID: one statement, matched in SQL.
  const [byCid] = await prisma.$queryRawUnsafe(
    `SELECT COUNT(*) AS n, COALESCE(SUM(rating = ${DEFAULT_RATING}), 0) AS r FROM listings WHERE ${MATCH}`,
  );
  console.log(`invented counts on rows with a CID: ${Number(byCid.n)} (${Number(byCid.r)} also rated ${DEFAULT_RATING})`);

  // 2. Rows whose CID was blanked later: recognised through the export's CID.
  let original = null;
  try {
    const rows = JSON.parse(await readFile(EXPORT_FILE, 'utf8'));
    original = new Map(rows.map((r) => [r.id, String(r.google_cid ?? r.googleCid ?? '')]));
  } catch {
    console.log(`no export at ${EXPORT_FILE}: rows without a CID cannot be checked`);
  }
  const blank = await prisma.listing.findMany({
    where: { OR: [{ googleCid: null }, { googleCid: '' }], reviewCount: { gt: 0 } },
    select: { id: true, reviewCount: true, rating: true },
  });
  const viaExport = original
    ? blank.filter((r) => original.get(r.id) && r.reviewCount === syntheticCount(original.get(r.id)))
    : [];
  console.log(
    `rows without a CID that have a count: ${blank.length}; invented per the export: ${viaExport.length}` +
      ` (${viaExport.filter((r) => r.rating === DEFAULT_RATING).length} also rated ${DEFAULT_RATING})`,
  );

  if (!APPLY) {
    console.log('Nothing written. Re-run with --apply to clear them.');
    return;
  }

  const n1 = await prisma.$executeRawUnsafe(
    `UPDATE listings SET rating = IF(rating = ${DEFAULT_RATING}, 0, rating), reviewCount = 0, updatedAt = NOW(3) WHERE ${MATCH}`,
  );
  let n2 = 0;
  for (const rated of [true, false]) {
    const ids = viaExport.filter((r) => (r.rating === DEFAULT_RATING) === rated).map((r) => r.id);
    for (let i = 0; i < ids.length; i += BATCH) {
      const res = await prisma.listing.updateMany({
        where: { id: { in: ids.slice(i, i + BATCH) } },
        data: rated ? { reviewCount: 0, rating: 0 } : { reviewCount: 0 },
      });
      n2 += res.count;
    }
  }
  console.log(`cleared: ${n1} rows by CID, ${n2} rows via the export. Restart the API to reload the index (instances running the listings sync pick it up on their own).`);
}

main()
  .catch((err) => {
    console.error(err);
    process.exitCode = 1;
  })
  .finally(() => prisma.$disconnect());
